libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3763,13 +3763,27 @@ bool TypeChecker::visit(Identifier const& _identifier)
 				"\"suicide\" has been deprecated in favour of \"selfdestruct\"."
 			);
 		else if (_identifier.name() == "selfdestruct" && fType->kind() == FunctionType::Kind::Selfdestruct)
-			m_errorReporter.warning(
-				5159_error,
-				_identifier.location(),
-				"\"selfdestruct\" has been deprecated. "
-				"The underlying opcode will eventually undergo breaking changes, "
-				"and its use is not recommended."
-			);
+		{
+			if (m_evmVersion <= langutil::EVMVersion::shanghai())
+				m_errorReporter.warning(
+					5159_error,
+					_identifier.location(),
+					"\"selfdestruct\" has been deprecated. "
+					"Since the VM version Cancun, the underlying opcode has changed its behavior in accordance with EIP-6780, "
+					"and its use is not recommended."
+				);
+			else
+				m_errorReporter.warning(
+					1249_error,
+					_identifier.location(),
+					"\"selfdestruct\" has been deprecated. "
+					"Since the VM version Cancun, \"selfdestruct\" functionality changed as defined by EIP-6780. "
+					"The new functionality only transfers all Ether in the account to the beneficiary. "
+					"However, the previous behavior is preserved when \"selfdestruct\" is called in the same transaction "
+					"in which a contract was created. See https://eips.ethereum.org/EIPS/eip-6780 for more information. "
+					"The use of \"selfdestruct\" is still not recommended."
```

Comment:
I'm not sure we need to split it into two versions. This might give an impression that changes only apply to contracts compiled for Cancun and if you compile for Shanghai, you still get the old behavior, but this is not the case. On mainnet all contracts will get the new behavior once the fork happens. You will only get the old behavior if you deploy to another network that did not implement `EIP-6780`.